---
title: "Phybion_ShortReads_final"
author: "Ahmed Hassan"
output: html_document
date: "2024-11-04"
---

#Loading the libraries

```{r}
library("tximeta")
library("SummarizedExperiment")
library("org.Hs.eg.db")
library("macrophage")
library("DESeq2")
library("clusterProfiler")
library("org.Hs.eg.db")
library("enrichplot")
library("mosdef")
library("GeneTonic")
library("IHW")
library("gplots")
library("genefilter")
library("rtracklayer")
library("xtable")
library("GO.db")
library("goseq")
library("GenomicFeatures")
library("ReactomePA")
library("dplyr")
library("NMF")
library("pheatmap")
library("GOplot")
library("knitr")
library("pcaExplorer")
library("DT")
library("ideal")
library("edgeR")
library("hexbin")
library("latticeExtra")
library("vsn")
library("gplots")
library("RColorBrewer")
library("topGO")
library("BIUMmisc")
library("biomaRt")
library("pathview")
library("VennDiagram")
library("GSEABase")
library("msigdbr")
library("ConsensusClusterPlus")
library("visNetwork")
library("igraph")
library("ggraph")
library("decoupleR")
library("OmnipathR")
library("quantiseqr")
library("immunedeconv")
library("tibble")
library("tidyr")
```


#Exploratory Data Analysis (EDA)

```{r}
# Loading the data

gse <- readRDS("gse_phybion_ShortReads.RDS")


dds <- DESeqDataSet(gse, design = ~ Donors + Dose)

dds <- dds[, c(
  "0Gy-2hrs-F", "0Gy-2hrs-A", "0Gy-2hrs-S", "0Gy-6hrs-F", "0Gy-6hrs-A",
  "0Gy-6hrs-S", "0.5Gy-2hrs-F", "0.5Gy-2hrs-A", "0.5Gy-2hrs-S", "0.5Gy-6hrs-F",
  "0.5Gy-6hrs-A", "0.5Gy-6hrs-S", "1Gy-2hrs-F", "1Gy-2hrs-A", "1Gy-2hrs-S",
  "1Gy-6hrs-F", "1Gy-6hrs-A", "1Gy-6hrs-S", "2Gy-2hrs-F", "2Gy-2hrs-A",
  "2Gy-2hrs-S", "2Gy-6hrs-F", "2Gy-6hrs-A", "2Gy-6hrs-S", "4Gy-2hrs-F",
  "4Gy-2hrs-A", "4Gy-2hrs-S", "4Gy-6hrs-F", "4Gy-6hrs-A", "4Gy-6hrs-S"
)]

dds <- dds_gencode_to_ensembl(dds)

dds_t1 <- dds[, dds$Time_point == "2hrs"]

dds_t2 <- dds[, dds$Time_point == "6hrs"]

dds$Dose <- factor(dds$Dose)

dds$Dose <- relevel(dds$Dose, ref = "0Gy")

dds$Time_point <- factor(dds$Time_point)

dds$Time_point <- relevel(dds$Time_point, ref = "2hrs")

dds_t1$Dose <- relevel(dds_t1$Dose, "0Gy")

dds_t2$Dose <- relevel(dds_t2$Dose, "0Gy")

dds_0Gy <- dds[, dds$Dose == "0Gy"]
```

## Variance stablizing transformation

```{r}
vsd_phybion <- vst(dds)
vsd_phybion_t1 <- vst(dds_t1)
vsd_phybion_t2 <- vst(dds_t2)


vsd_corrected <- vsd_phybion
assay(vsd_corrected) <- limma::removeBatchEffect(assay(vsd_phybion), batch = vsd_phybion$Donors)

vsd_t1_corrected <- vsd_phybion_t1
assay(vsd_t1_corrected) <- limma::removeBatchEffect(assay(vsd_phybion_t1), batch = vsd_phybion_t1$Donors)

vsd_t2_corrected <- vsd_phybion_t2
assay(vsd_t2_corrected) <- limma::removeBatchEffect(assay(vsd_phybion_t2), batch = vsd_phybion_t2$Donors)
```

## checking read coverage over the samples

```{r}
plot_totcounts(dds, group = "Dose")
plot_totcounts(dds, group = "Donors")
plot_totcounts(dds, group = "Time_point")

## PCA plot

pca_plot <- plotPCA(vsd_phybion, intgroup = c("Donors", "Time_point"), returnData = TRUE)

percentVar <- round(100 * attr(pca_plot, "percentVar"))

ggplot(pca_plot, aes(x = PC1, y = PC2, color = Donors, shape = Time_point)) +
  geom_point(size = 3) +
  scale_shape_manual(values = c(16, 17)) + # Adjust the number of shapes if necessary
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  theme_minimal()


# clustering samples

sampleDists <- dist(t(assay(vsd_corrected)))

sampleDistMatrix <- as.matrix(sampleDists)

colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)


# Extract time point from sample names
time_point <- ifelse(grepl("2hrs", colnames(sampleDistMatrix)), "2hrs", "6hrs")

# Extract sex from sample names:
sex <- ifelse(grepl("F$", colnames(sampleDistMatrix)), "Female", "Male")

# Extract Dose from sample names:
dose <- sub("-.*", "", colnames(sampleDistMatrix))

# Create a data frame for the column annotation
column_annotation <- data.frame(Time_Point = time_point, Sex = sex, Dose = dose)

# Set row names of annotation to match the column names of sampleDistMatrix
rownames(column_annotation) <- colnames(sampleDistMatrix)

pheatmap(sampleDistMatrix,
  clustering_distance_rows = sampleDists,
  clustering_distance_cols = sampleDists, annotation_col = column_annotation,
  col = colors
)

# clustering Genes

topVarGenes <- head(order(rowVars(assay(vsd_corrected)), decreasing = TRUE), 100)

mat <- assay(vsd_corrected)[topVarGenes, ]

mat <- mat - rowMeans(mat)

anno <- as.data.frame(colData(vsd_corrected)[, c("Sex", "Time_point", "Donors")])

genes <- rownames(mat)

genes <- mapIds(org.Hs.eg.db,
  keys = genes,
  column = "SYMBOL",
  keytype = "ENSEMBL",
  multiVals = "first"
)

rownames(mat) <- genes

pheatmap(mat, annotation_col = anno, cluster_cols = F)
```

# Annotations


```{r}
# Annotations for the whole experiment

anno_df <- get_annotation_orgdb(
  dds = dds,
  orgdb_species = "org.Hs.eg.db",
  idtype = "ENSEMBL"
)

mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl")

anns <- getBM(
  attributes = c("ensembl_gene_id", "external_gene_name", "description"),
  filters = "ensembl_gene_id",
  values = rownames(dds),
  mart = mart
)
```

# Differential expression analysis

## Merging all DEA into one big list

```{r}
# Time point1

extended_anno <- fortify_annotations(
  anno_df_1 = anno_df,
  anno_df_2 = anns,
  dds = dds
)

# All doses VS 0 Gy comparison

dds_t1 <- DESeq(dds_t1)

resultsNames(dds_t1)


resuSet <-
  create_DEresults(
    resuSet = NULL,
    dds_obj = dds_t1,
    contrast_name = "Dose_4Gy_vs_0Gy",
    FDR = 0.05,
    extended_anno_df = extended_anno$anno_df,
    species = "Homo_sapiens",
    name_res_entry = "t1_4Gy_vs_0Gy"
  )

resuSet <-
  create_DEresults(
    resuSet = resuSet,
    dds_obj = dds_t1,
    contrast_name = "Dose_2Gy_vs_0Gy",
    FDR = 0.05,
    extended_anno_df = extended_anno$anno_df,
    species = "Homo_sapiens",
    name_res_entry = "t1_2Gy_vs_0Gy"
  )


resuSet <-
  create_DEresults(
    resuSet = resuSet,
    dds_obj = dds_t1,
    contrast_name = "Dose_0.5Gy_vs_0Gy",
    FDR = 0.05,
    extended_anno_df = extended_anno$anno_df,
    species = "Homo_sapiens",
    name_res_entry = "t1_0.5Gy_vs_0Gy"
  )



resuSet <-
  create_DEresults(
    resuSet = resuSet,
    dds_obj = dds_t1,
    contrast_name = "Dose_1Gy_vs_0Gy",
    FDR = 0.05,
    extended_anno_df = extended_anno$anno_df,
    species = "Homo_sapiens",
    name_res_entry = "t1_1Gy_vs_0Gy"
  )

# Time point2


# All doses VS 0 Gy comparison

dds_t2 <- DESeq(dds_t2)

resultsNames(dds_t2)


resuSet <-
  create_DEresults(
    resuSet = resuSet,
    dds_obj = dds_t2,
    contrast_name = "Dose_4Gy_vs_0Gy",
    FDR = 0.05,
    extended_anno_df = extended_anno$anno_df,
    species = "Homo_sapiens",
    name_res_entry = "t2_4Gy_vs_0Gy"
  )

resuSet <-
  create_DEresults(
    resuSet = resuSet,
    dds_obj = dds_t2,
    contrast_name = "Dose_2Gy_vs_0Gy",
    FDR = 0.05,
    extended_anno_df = extended_anno$anno_df,
    species = "Homo_sapiens",
    name_res_entry = "t2_2Gy_vs_0Gy"
  )

resuSet <-
  create_DEresults(
    resuSet = resuSet,
    dds_obj = dds_t2,
    contrast_name = "Dose_1Gy_vs_0Gy",
    FDR = 0.05,
    extended_anno_df = extended_anno$anno_df,
    species = "Homo_sapiens",
    name_res_entry = "t2_1Gy_vs_0Gy"
  )


resuSet <-
  create_DEresults(
    resuSet = resuSet,
    dds_obj = dds_t2,
    contrast_name = "Dose_0.5Gy_vs_0Gy",
    FDR = 0.05,
    extended_anno_df = extended_anno$anno_df,
    species = "Homo_sapiens",
    name_res_entry = "t2_0.5Gy_vs_0Gy"
  )
```

## effect of incubation on gene expression

```{r}
design(dds_0Gy) <- ~ Donors + Time_point
dds_0Gy <- DESeq(dds_0Gy)
resultsNames(dds_0Gy)

res_0Gy_t2_vs_t1 <- results(dds_0Gy, name = "Time_point_6hrs_vs_2hrs", alpha = 0.05)

summary(res_0Gy_t2_vs_t1)

res_0Gy_t2_vs_t1 <- deresult_to_df(res_0Gy_t2_vs_t1, FDR = 0.05)

genes <- rownames(res_0Gy_t2_vs_t1)

genes <- mapIds(
  org.Hs.eg.db,
  keys = genes,
  column = "SYMBOL",
  keytype = "ENSEMBL",
  multiVals = "first"
)


res_0Gy_t2_vs_t1$SYMBOL <- genes

res_enrich_0Gy_t2_vs_t1 <- enrichGO(
  gene = res_0Gy_t2_vs_t1$SYMBOL,
  universe = geneUniverseExpr,
  keyType = "SYMBOL",
  OrgDb = org.Hs.eg.db,
  ont = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = FALSE
)

writexl::write_xlsx(res_enrich_0Gy_t2_vs_t1@result, "res_enrich_0Gy_t2_vs_t1.xlsx")

writexl::write_xlsx(res_0Gy_t2_vs_t1, "res_0Gy_t2_vs_t1.xlsx")
```

## Volcano plot

```{r}

volcPlot <- de_volcano(
  res_de = resuSet$t1_4Gy_vs_0Gy$res_DESeq,
  mapping = "org.Hs.eg.db",
  labeled_genes = 50,
  logfc_cutoff = 1
)
volcPlot

```

## Creating gene plotting function

```{r}
plot_gene <- function (dds, gene, intgroup = "condition", assay = "counts", 
                       annotation_obj = NULL, normalized = TRUE, transform = TRUE, 
                       labels_display = FALSE, labels_repel = TRUE, plot_type = "auto", 
                       return_data = FALSE, color_by = "Donors", gtl = NULL) 
{
  plot_type <- match.arg(plot_type, c("auto", "jitteronly", 
                                      "boxplot", "violin", "sina"))
  
  if (!is.null(gtl)) {
    checkup_gtl(gtl)
    dds <- gtl$dds
    res_de <- gtl$res_de
    res_enrich <- gtl$res_enrich
    annotation_obj <- gtl$annotation_obj
  }

  if (!(all(intgroup %in% colnames(colData(dds))))) {
    stop("`intgroup` not found in the colData slot of the dds object", 
         "\nPlease specify one of the following: \n", paste0(colnames(colData(dds)), 
                                                             collapse = ", "))
  }
  
  df <- get_expression_values(dds = dds, gene = gene, intgroup = intgroup, 
                              assay = assay, normalized = normalized)
  df$sample_id <- rownames(df)
  
  # Explicitly add the Donors column from colData(dds) if it exists
  if ("Donors" %in% colnames(colData(dds))) {
    df$Donors <- colData(dds)[, "Donors"]
  } else {
    stop("The `Donors` column is missing in the colData of the dds object.")
  }
  
  if (!is.null(annotation_obj)) {
    genesymbol <- annotation_obj$gene_name[match(gene, annotation_obj$gene_id)]
  } else {
    genesymbol <- ""
  }
  
  onlyfactors <- df[, match(intgroup, colnames(df))]
  df$plotby <- interaction(onlyfactors)
  min_by_groups <- min(table(df$plotby))
  
  if (return_data) {
    return(df)
  }

  p <- ggplot(df, aes(x = .data$plotby, y = .data$exp_value, col = .data[[color_by]])) + 
    scale_x_discrete(name = "") + 
    scale_color_discrete(name = "Experimental\ngroup") + 
    theme_bw()

  # Add error bars for standard deviation
  #p <- p + stat_summary(fun.data = mean_sdl, fun.args = list(mult = 1), geom = "errorbar", width = 0.2, color = "black")

  jit_pos <- position_jitter(width = 0.2, height = 0, seed = 42)
  
  if (plot_type == "jitteronly" || (plot_type == "auto" & min_by_groups <= 3)) {
    p <- p + geom_point(position = jit_pos)
  } else if (plot_type == "boxplot" || (plot_type == "auto" & (min_by_groups > 3 & min_by_groups < 10))) {
    p <- p + geom_boxplot(outlier.shape = NA) + geom_point(position = jit_pos)
  } else if (plot_type == "violin" || (plot_type == "auto" & (min_by_groups >= 11 & min_by_groups < 40))) {
    p <- p + geom_violin() + geom_point(position = jit_pos) + 
      stat_summary(fun = median, fun.min = median, fun.max = median, geom = "crossbar", width = 0.3)
  } else if (plot_type == "sina" || (plot_type == "auto" & (min_by_groups >= 40))) {
    p <- p + ggforce::geom_sina() + stat_summary(fun = median, 
                                                 fun.min = median, fun.max = median, geom = "crossbar", width = 0.3)
  }

  # Remove sample labels from plot by setting `labels_display = FALSE`
  if (labels_display) {
    if (labels_repel) {
      p <- p + ggrepel::geom_text_repel(aes(label = .data$sample_id), 
                                        min.segment.length = 0, position = jit_pos)
    } else {
      p <- p + geom_text(aes(label = .data$sample_id), hjust = -0.1, vjust = 0.1, position = jit_pos)
    }
  }
  
  # Add grey line representing the mean across all groups
  p <- p + stat_summary(fun = mean, geom = "line", aes(group = 1), 
                        color = "grey80", linewidth = 0.5)

  y_label <- if (assay == "counts" & normalized) {
    "Normalized counts"
  } else if (assay == "counts" & !normalized) {
    "Counts"
  } else if (assay == "abundance") {
    "TPM - Transcripts Per Million"
  } else {
    assay
  }
  
  if (transform) {
    p <- p + scale_y_log10(name = paste0(y_label, " (log10 scale)"))
  } else {
    p <- p + scale_y_continuous(name = y_label)
  }

  if (!is.null(annotation_obj)) {
    p <- p + labs(title = paste0(genesymbol, " - ", gene))
  } else {
    p <- p + labs(title = paste0(gene))
  }

  return(p)
}


```

## Checking the gene pattern across doses

```{r}

plot_gene(
  dds= dds_t1,
  gene = "ENSG00000169429",
  intgroup = "Dose",
  annotation_obj = anno_df
)
```

## since we spotted time effect on non-irradiated samples, DEA to capture radiation-only late effect would be as such

```{r}
design(dds) <- ~ Donors + Time_point + Dose + Time_point:Dose

dds <- DESeq(dds)

resultsNames(dds)


# 4Gy t2 Vs 4Gy t1

res_4Gy_time2_vs_time1 <- results(dds,
  name = "Time_point6hrs.Dose4Gy",
  test = "Wald",
  alpha = 0.05
)

res_shrink_4Gy_time2_vs_time1 <- lfcShrink(dds,
  res = res_4Gy_time2_vs_time1,
  coef = "Time_point6hrs.Dose4Gy",
  type = "apeglm"
)

summary(res_shrink_4Gy_time2_vs_time1)

res_shrink_4Gy_time2_vs_time1 <- deresult_to_df(res_shrink_4Gy_time2_vs_time1, FDR = 0.05)

# 2Gy t2 Vs 2Gy t1

res_2Gy_time2_vs_time1 <- results(dds,
  name = "Time_point6hrs.Dose2Gy",
  test = "Wald",
  alpha = 0.05
)

res_shrink_2Gy_time2_vs_time1 <- lfcShrink(dds,
  res = res_2Gy_time2_vs_time1,
  coef = "Time_point6hrs.Dose2Gy",
  type = "apeglm"
)

summary(res_shrink_2Gy_time2_vs_time1)

res_shrink_2Gy_time2_vs_time1 <- deresult_to_df(res_shrink_2Gy_time2_vs_time1, FDR = 0.05)

# 1Gy t2 Vs 1Gy t1

res_1Gy_time2_vs_time1 <- results(dds,
  name = "Time_point6hrs.Dose1Gy",
  test = "Wald",
  alpha = 0.05
)

res_shrink_2Gy_time2_vs_time1 <- lfcShrink(dds,
  res = res_1Gy_time2_vs_time1,
  coef = "Time_point6hrs.Dose1Gy",
  type = "apeglm"
)

summary(res_shrink_2Gy_time2_vs_time1)

res_shrink_2Gy_time2_vs_time1 <- deresult_to_df(res_shrink_2Gy_time2_vs_time1, FDR = 0.05)

# 0.5Gy t2 Vs 0.5Gy t1

res_0.5Gy_time2_vs_time1 <- results(dds,
  name = "Time_point6hrs.Dose0.5Gy",
  test = "Wald",
  alpha = 0.05
)

res_shrink_0.5Gy_time2_vs_time1 <- lfcShrink(dds,
  res = res_0.5Gy_time2_vs_time1,
  coef = "Time_point6hrs.Dose0.5Gy",
  type = "apeglm"
)

summary(res_shrink_0.5Gy_time2_vs_time1)

res_shrink_0.5Gy_time2_vs_time1 <- deresult_to_df(res_shrink_0.5Gy_time2_vs_time1, FDR = 0.05)
```

# Functional Enrichment Analysis

```{r}
expressedInAssay <- (rowSums(assay(dds)) > 0)
geneUniverseExprENS <- rownames(dds)[expressedInAssay]
geneUniverseExpr <- anno_df$gene_name[match(geneUniverseExprENS, anno_df$gene_id)]
```

## Enrichment with ClusterProfiler

```{r}
# Enrichment with clusterProfiler

for (i in names(resuSet)) {
  message(i)
  if (nrow(resuSet[[i]][["tbl_res_DE"]]) > 0) {
    resuSet[[i]][["clupro_tbl"]] <- run_cluPro(
      de_genes = resuSet[[i]][["tbl_res_DE"]]$gene_name,
      res_de = resuSet[[i]][["res_DESeq"]],
      de_container = dds,
      keyType = "SYMBOL",
      mapping = "org.Hs.eg.db",
      ont = "BP",
      pAdjustMethod = "BH",
      bg_genes = geneUniverseExpr,
      FDR_threshold = 0.05,
    )
  }
}
```

## Creating GeneTonic objects

```{r}
gtl_t1_0.5Gy <- GeneTonicList(
  dds = dds,
  res_de = resuSet$t1_0.5Gy_vs_0Gy$res_DESeq,
  res_enrich = shake_enrichResult(resuSet$t1_0.5Gy_vs_0Gy$clupro_tbl),
  annotation_obj = anno_df
)

gtl_t1_1Gy <- GeneTonicList(
  dds = dds,
  res_de = resuSet$t1_1Gy_vs_0Gy$res_DESeq,
  res_enrich = shake_enrichResult(resuSet$t1_1Gy_vs_0Gy$clupro_tbl),
  annotation_obj = anno_df
)

gtl_t1_2Gy <- GeneTonicList(
  dds = dds,
  res_de = resuSet$t1_2Gy_vs_0Gy$res_DESeq,
  res_enrich = shake_enrichResult(resuSet$t1_2Gy_vs_0Gy$clupro_tbl),
  annotation_obj = anno_df
)

gtl_t1_4Gy <- GeneTonicList(
  dds = dds,
  res_de = resuSet$t1_4Gy_vs_0Gy$res_DESeq,
  res_enrich = shake_enrichResult(resuSet$t1_4Gy_vs_0Gy$clupro_tbl),
  annotation_obj = anno_df
)

gtl_t2_0.5Gy <- GeneTonicList(
  dds = dds,
  res_de = resuSet$t2_0.5Gy_vs_0Gy$res_DESeq,
  res_enrich = shake_enrichResult(resuSet$t2_0.5Gy_vs_0Gy$clupro_tbl),
  annotation_obj = anno_df
)

gtl_t2_1Gy <- GeneTonicList(
  dds = dds,
  res_de = resuSet$t2_1Gy_vs_0Gy$res_DESeq,
  res_enrich = shake_enrichResult(resuSet$t2_1Gy_vs_0Gy$clupro_tbl),
  annotation_obj = anno_df
)

gtl_t2_2Gy <- GeneTonicList(
  dds = dds,
  res_de = resuSet$t2_2Gy_vs_0Gy$res_DESeq,
  res_enrich = shake_enrichResult(resuSet$t2_2Gy_vs_0Gy$clupro_tbl),
  annotation_obj = anno_df
)

gtl_t2_4Gy <- GeneTonicList(
  dds = dds,
  res_de = resuSet$t2_4Gy_vs_0Gy$res_DESeq,
  res_enrich = shake_enrichResult(resuSet$t2_4Gy_vs_0Gy$clupro_tbl),
  annotation_obj = anno_df
)
```

## Running GeneTonic

```{r}
GeneTonic(gtl = gtl_t1_4Gy)
```

### GeneTonic example plots

```{r}
# GeneTonic Pathway-associated DEGs Heatmap

gs_heatmap(
  se = vsd_corrected,
  gtl = gtl_t2_0.5Gy,
  geneset_id = "GO:0008630", anno_col_info = c("Dose", "Time_point", "Sex"), cluster_columns = FALSE
)


# GeneTonic Geneset Heatmap t1_0.5Gy_vs_0Gy

anno_t2 <- as.data.frame(colData(vsd_phybion_t2)[, c("Sex", "Time_point", "Donors", "Dose")])
anno_t1 <- as.data.frame(colData(vsd_phybion_t1)[, c("Sex", "Time_point", "Donors", "Dose")])
anno_full <- as.data.frame(colData(vsd_phybion)[, c("Sex", "Time_point", "Donors", "Dose")])


vsd_corrected_ordred <- vsd_corrected[, c(1, 2, 3, 7, 8, 9, 13, 14, 15, 19, 20, 21, 25, 26, 27, 4, 5, 6, 10, 11, 12, 16, 17, 18, 22, 23, 24, 28, 29, 30)]

scored_mat_t2_4Gy_vs_0Gy <- GeneTonic::gs_scores(se = vsd_t2_corrected, gtl = gtl_t2_4Gy)

column_annotation <- HeatmapAnnotation(
  Sex = anno_t2$Sex,
  Dose = anno_t2$Dose,
  col = anno_colors
)


Heatmap(scored_mat_t2_4Gy_vs_0Gy[1:30, ],
  name = "Expression",
  top_annotation = column_annotation,
  cluster_columns = F, # Optional: Cluster columns
  cluster_rows = F, # Optional: Cluster rows
  show_row_names = TRUE,
  show_column_names = TRUE, row_dend_side = "left",
  row_names_side = "left",
  row_names_gp = gpar(fontsize = 4),
  width = unit(12, "cm"),
  heatmap_height = unit(12, "cm"),
  column_names_gp = gpar(fontsize = 5)
)

# OR

GeneTonic::gs_scoresheat(scored_mat_t2_4Gy_vs_0Gy, n_gs = 30, cluster_cols = FALSE)
```

## GSEA

```{r}
## background genes

background_genes <- resuSet$t2_4Gy_vs_0Gy$tbl_res_all %>%
  as.data.frame() %>%
  filter(baseMean != 0) %>%
  tibble::rownames_to_column(var = "gene") %>%
  pull(gene)


res_df <- resuSet$t2_4Gy_vs_0Gy$tbl_res_all %>%
  as.data.frame() %>%
  filter(baseMean != 0) %>%
  tibble::rownames_to_column(var = "gene")

background_genes_map <- bitr(
  geneID = background_genes,
  fromType = "ENSEMBL",
  toType = "ENTREZID",
  OrgDb = "org.Hs.eg.db"
)


m_df <- msigdbr(species = "Homo sapiens")

head(m_df)

m_t2g <- msigdbr(species = "Homo sapiens", category = "H") %>%
  dplyr::select(gs_name, entrez_gene)


table(m_t2g$gs_name)
head(m_t2g)

### Gene set enrichment analysis

## rank all the genes by signed fold change * -log10pvalue.

res_df <- res_df %>%
  mutate(signed_rank_stats = sign(log2FoldChange) * -log10(pvalue)) %>%
  left_join(background_genes_map, by = c("gene" = "ENSEMBL")) %>%
  arrange(desc(signed_rank_stats))

gene_list <- res_df$signed_rank_stats
names(gene_list) <- res_df$ENTREZID

em2 <- GSEA(gene_list, TERM2GENE = m_t2g)

em2@result %>% View()

### visualization

p1 <- gseaplot(em2,
  geneSetID = "HALLMARK_TNFA_SIGNALING_VIA_NFKB",
  by = "runningScore", title = "HALLMARK_TNFA_SIGNALING_VIA_NFKB"
)

p2 <- gseaplot(em2,
  geneSetID = "HALLMARK_INFLAMMATORY_RESPONSE",
  by = "runningScore", title = "HALLMARK_INFLAMMATORY_RESPONSE"
)

p1 / p2
```

# Gene Set Variation Analysis

```{r}
exprs_mat <- assay(vsd_phybion)

genes <- rownames(exprs_mat)

genes <- mapIds(
  org.Hs.eg.db,
  keys = genes,
  column = "SYMBOL",
  keytype = "ENSEMBL",
  multiVals = "first"
)

rownames(exprs_mat) <- genes

goannot <- select(org.Hs.eg.db, keys = keys(org.Hs.eg.db, "SYMBOL"), columns = "GO", keytype = "SYMBOL")

genesbygo <- split(goannot$SYMBOL, goannot$GO)

huangPar <- gsvaParam(exprs_mat, genesbygo, minSize = 5, maxSize = 500)

esmicro <- gsva(huangPar)

# Applying Limma on GSVA matrix

# creating metadata

patient <- rep(c("p1", "p2", "p3"), 10)
dose <- c(rep("0Gy", 6), rep("0.5Gy", 6), rep("1Gy", 6), rep("2Gy", 6), rep("4Gy", 6))
time <- rep(c(rep("2hrs", 3), rep("6hrs", 3)), 5)
metadata <- data.frame(patient, dose, time)


f <- factor(paste(metadata$dose, metadata$time, sep = "_"))

design <- model.matrix(~ 0 + f)

patient <- as.factor(metadata$patient)

corfit <- duplicateCorrelation(esmicro, design, block = patient)

fit <- lmFit(esmicro, design, block = patient, correlation = corfit$consensus)

contrast.matrix <- makeContrasts(f4Gy_6hrs - f0Gy_6hrs, levels = design)

fit2 <- contrasts.fit(fit, contrast.matrix)

fit2 <- eBayes(fit2)

results_gsva <- topTable(fit2, coef = 1, number = nrow(esmicro), adjust = "BH")

results_f_gsva <- decideTests(fit2, p.value = 0.05)

summary(results_f_gsva)
```

# DecoupleR-- Transcription Factors inference

## Filtering

```{r}
# extracting the normalized matrix

decoupler_counts <- assay(vsd_phybion)

decoupler_genes <- rownames(decoupler_counts)

decoupler_genes <- mapIds(
  org.Hs.eg.db,
  keys = decoupler_genes,
  column = "SYMBOL",
  keytype = "ENSEMBL",
  multiVals = "first"
)

decoupler_counts <- as.data.frame(decoupler_counts)

decoupler_counts$SYMBOL <- decoupler_genes
decoupler_counts$id <- rownames(decoupler_counts)

# extracting the raw matrix

dds_mat <- counts(dds)

dds_genes <- rownames(dds_mat)

dds_genes <- mapIds(
  org.Hs.eg.db,
  keys = dds_genes,
  column = "SYMBOL",
  keytype = "ENSEMBL",
  multiVals = "first"
)

dds_mat <- as.data.frame(dds_mat)

Vars_raw <- rowVars(dds_mat)

dds_mat$SYMBOL <- dds_genes
dds_mat$id <- rownames(dds_mat)
dds_mat$Vars_raw <- Vars_raw


dds_counts_dup <- dds_mat$SYMBOL[duplicated(dds_mat$SYMBOL)]

dds_counts_dup <- dds_mat[dds_mat$SYMBOL %in% dds_counts_dup, c("SYMBOL", "id")]

dds_counts_dup <- dds_mat[dds_counts_dup$id, ]

# extracting the duplicates that have higher variablity

dds_counts_dup <- dds_counts_dup %>% rownames_to_column("Ensembl_ID")


filtered_ddd <- dds_counts_dup %>%
  group_by(SYMBOL) %>%
  slice_max(order_by = Vars_raw, n = 1) %>%
  ungroup()

# Convert back to rownames

filtered_ddd <- filtered_ddd %>% column_to_rownames("Ensembl_ID")

dds_counts_dup <- dds_counts_dup %>% column_to_rownames("Ensembl_ID")

#  extracting the duplicates that have lower variablity to filter them out

ddd_low_var <- as.data.frame(dds_counts_dup[!rownames(dds_counts_dup) %in% rownames(filtered_ddd), ])

# filter the low variable duplicates from normalized matrix

decoupler_counts <- decoupler_counts[!rownames(decoupler_counts) %in% rownames(ddd_low_var), ]

## all the remaining 33 duplicates are genes with 0 raw count

decoupler_counts <- decoupler_counts[!duplicated(decoupler_counts$SYMBOL), ]


decoupler_counts <- na.omit(decoupler_counts)
rownames(decoupler_counts) <- decoupler_counts$SYMBOL
decoupler_counts <- decoupler_counts[, -c(31, 32)]

# filter the low variable duplicates from raw matrix

dds_mat <- dds_mat[!rownames(dds_mat) %in% rownames(ddd_low_var), ]

dds_mat <- na.omit(dds_mat)

dds_mat$rowSums <- rowSums(dds_mat[, 1:30])

dds_mat <- dds_mat[dds_mat$rowSums > 0, ]

rownames(dds_mat) <- dds_mat$SYMBOL

dds_mat <- dds_mat[, -c(31, 32, 33, 34)]
```

## Time point 1


```{r}
# t1_0.5Gy
res_t1_0.5Gy_vs_0Gy <- resuSet$t1_0.5Gy_vs_0Gy$res_DESeq

res_t1_0.5Gy_vs_0Gy <- deresult_to_df(res_t1_0.5Gy_vs_0Gy)

res_t1_0.5Gy_vs_0Gy <- res_t1_0.5Gy_vs_0Gy[!rownames(res_t1_0.5Gy_vs_0Gy) %in% rownames(ddd_low_var), ]

res_t1_0.5Gy_vs_0Gy <- na.omit(res_t1_0.5Gy_vs_0Gy)

# Filter out rows with any blank cells
res_t1_0.5Gy_vs_0Gy <- res_t1_0.5Gy_vs_0Gy[!apply(res_t1_0.5Gy_vs_0Gy == "", 1, any), ]

rownames(res_t1_0.5Gy_vs_0Gy) <- res_t1_0.5Gy_vs_0Gy$SYMBOL

res_t1_0.5Gy_vs_0Gy <- res_t1_0.5Gy_vs_0Gy %>%
  mutate(stat = sign(log2FoldChange) * -log10(pvalue)) %>%
  arrange(desc(stat))



# t1_1Gy

res_t1_1Gy_vs_0Gy <- resuSet$t1_1Gy_vs_0Gy$res_DESeq

res_t1_1Gy_vs_0Gy <- deresult_to_df(res_t1_1Gy_vs_0Gy)

res_t1_1Gy_vs_0Gy <- res_t1_1Gy_vs_0Gy[!rownames(res_t1_1Gy_vs_0Gy) %in% rownames(ddd_low_var), ]

res_t1_1Gy_vs_0Gy <- na.omit(res_t1_1Gy_vs_0Gy)
# Filter out rows with any blank cells
res_t1_1Gy_vs_0Gy <- res_t1_1Gy_vs_0Gy[!apply(res_t1_1Gy_vs_0Gy == "", 1, any), ]

rownames(res_t1_1Gy_vs_0Gy) <- res_t1_1Gy_vs_0Gy$SYMBOL

res_t1_1Gy_vs_0Gy <- res_t1_1Gy_vs_0Gy %>%
  mutate(stat = sign(log2FoldChange) * -log10(pvalue)) %>%
  arrange(desc(stat))


# t1_2Gy

res_t1_2Gy_vs_0Gy <- resuSet$t1_2Gy_vs_0Gy$res_DESeq

res_t1_2Gy_vs_0Gy <- deresult_to_df(res_t1_2Gy_vs_0Gy)

res_t1_2Gy_vs_0Gy <- res_t1_2Gy_vs_0Gy[!rownames(res_t1_2Gy_vs_0Gy) %in% rownames(ddd_low_var), ]


res_t1_2Gy_vs_0Gy <- na.omit(res_t1_2Gy_vs_0Gy)
# Filter out rows with any blank cells
res_t1_2Gy_vs_0Gy <- res_t1_2Gy_vs_0Gy[!apply(res_t1_2Gy_vs_0Gy == "", 1, any), ]

rownames(res_t1_2Gy_vs_0Gy) <- res_t1_2Gy_vs_0Gy$SYMBOL

res_t1_2Gy_vs_0Gy <- res_t1_2Gy_vs_0Gy %>%
  mutate(stat = sign(log2FoldChange) * -log10(pvalue)) %>%
  arrange(desc(stat))

# t1_4Gy

res_t1_4Gy_vs_0Gy <- resuSet$t1_4Gy_vs_0Gy$res_DESeq

res_t1_4Gy_vs_0Gy <- deresult_to_df(res_t1_4Gy_vs_0Gy)

res_t1_4Gy_vs_0Gy <- res_t1_4Gy_vs_0Gy[!rownames(res_t1_4Gy_vs_0Gy) %in% rownames(ddd_low_var), ]

res_t1_4Gy_vs_0Gy <- na.omit(res_t1_4Gy_vs_0Gy)
# Filter out rows with any blank cells
res_t1_4Gy_vs_0Gy <- res_t1_4Gy_vs_0Gy[!apply(res_t1_4Gy_vs_0Gy == "", 1, any), ]

rownames(res_t1_4Gy_vs_0Gy) <- res_t1_4Gy_vs_0Gy$SYMBOL

res_t1_4Gy_vs_0Gy <- res_t1_4Gy_vs_0Gy %>%
  mutate(stat = sign(log2FoldChange) * -log10(pvalue)) %>%
  arrange(desc(stat))
```

## Time point 2
```{r}
# t2_0.5Gy

res_t2_0.5Gy_vs_0Gy <- resuSet$t2_0.5Gy_vs_0Gy$res_DESeq

res_t2_0.5Gy_vs_0Gy <- deresult_to_df(res_t2_0.5Gy_vs_0Gy)

res_t2_0.5Gy_vs_0Gy <- res_t2_0.5Gy_vs_0Gy[!rownames(res_t2_0.5Gy_vs_0Gy) %in% rownames(ddd_low_var), ]

res_t2_0.5Gy_vs_0Gy <- na.omit(res_t2_0.5Gy_vs_0Gy)

# Filter out rows with any blank cells
res_t2_0.5Gy_vs_0Gy <- res_t2_0.5Gy_vs_0Gy[!apply(res_t2_0.5Gy_vs_0Gy == "", 1, any), ]

rownames(res_t2_0.5Gy_vs_0Gy) <- res_t2_0.5Gy_vs_0Gy$SYMBOL

res_t2_0.5Gy_vs_0Gy <- res_t2_0.5Gy_vs_0Gy %>%
  mutate(stat = sign(log2FoldChange) * -log10(pvalue)) %>%
  arrange(desc(stat))

# t2_1Gy

res_t2_1Gy_vs_0Gy <- resuSet$t2_1Gy_vs_0Gy$res_DESeq

res_t2_1Gy_vs_0Gy <- deresult_to_df(res_t2_1Gy_vs_0Gy)

res_t2_1Gy_vs_0Gy <- res_t2_1Gy_vs_0Gy[!rownames(res_t2_1Gy_vs_0Gy) %in% rownames(ddd_low_var), ]

res_t2_1Gy_vs_0Gy <- na.omit(res_t2_1Gy_vs_0Gy)
# Filter out rows with any blank cells
res_t2_1Gy_vs_0Gy <- res_t2_1Gy_vs_0Gy[!apply(res_t2_1Gy_vs_0Gy == "", 1, any), ]

rownames(res_t2_1Gy_vs_0Gy) <- res_t2_1Gy_vs_0Gy$SYMBOL

res_t2_1Gy_vs_0Gy <- res_t2_1Gy_vs_0Gy %>%
  mutate(stat = sign(log2FoldChange) * -log10(pvalue)) %>%
  arrange(desc(stat))

# t2_2Gy

res_t2_2Gy_vs_0Gy <- resuSet$t2_2Gy_vs_0Gy$res_DESeq

res_t2_2Gy_vs_0Gy <- deresult_to_df(res_t2_2Gy_vs_0Gy)

res_t2_2Gy_vs_0Gy <- res_t2_2Gy_vs_0Gy[!rownames(res_t2_2Gy_vs_0Gy) %in% rownames(ddd_low_var), ]

res_t2_2Gy_vs_0Gy <- na.omit(res_t2_2Gy_vs_0Gy)
# Filter out rows with any blank cells
res_t2_2Gy_vs_0Gy <- res_t2_2Gy_vs_0Gy[!apply(res_t2_2Gy_vs_0Gy == "", 1, any), ]

rownames(res_t2_2Gy_vs_0Gy) <- res_t2_2Gy_vs_0Gy$SYMBOL

res_t2_2Gy_vs_0Gy <- res_t2_2Gy_vs_0Gy %>%
  mutate(stat = sign(log2FoldChange) * -log10(pvalue)) %>%
  arrange(desc(stat))

# t2_4Gy

res_t2_4Gy_vs_0Gy <- resuSet$t2_4Gy_vs_0Gy$res_DESeq

res_t2_4Gy_vs_0Gy <- deresult_to_df(res_t2_4Gy_vs_0Gy)

res_t2_4Gy_vs_0Gy <- res_t2_4Gy_vs_0Gy[!rownames(res_t2_4Gy_vs_0Gy) %in% rownames(ddd_low_var), ]

res_t2_4Gy_vs_0Gy <- na.omit(res_t2_4Gy_vs_0Gy)
# Filter out rows with any blank cells
res_t2_4Gy_vs_0Gy <- res_t2_4Gy_vs_0Gy[!apply(res_t2_4Gy_vs_0Gy == "", 1, any), ]

rownames(res_t2_4Gy_vs_0Gy) <- res_t2_4Gy_vs_0Gy$SYMBOL

res_t2_4Gy_vs_0Gy <- res_t2_4Gy_vs_0Gy %>%
  mutate(stat = sign(log2FoldChange) * -log10(pvalue)) %>%
  arrange(desc(stat))
```

## Performing the analysis

```{r}
net <- get_collectri(organism = "human", split_complexes = FALSE)

sample_acts_t1_0.5Gy <- run_ulm(
  mat = res_t1_0.5Gy_vs_0Gy[, "stat", drop = FALSE],
  net = net,
  .source = "source",
  .target = "target",
  .mor = "mor"
)

sig_sample_acts_t1_0.5Gy <- sample_acts_t1_0.5Gy[sample_acts_t1_0.5Gy$p_value <= 0.05, ]

sample_acts_t1_1Gy <- run_ulm(
  mat = res_t1_1Gy_vs_0Gy[, "stat", drop = FALSE],
  net = net,
  .source = "source",
  .target = "target",
  .mor = "mor"
)

sig_sample_acts_t1_1Gy <- sample_acts_t1_1Gy[sample_acts_t1_1Gy$p_value <= 0.05, ]

sample_acts_t1_2Gy <- run_ulm(
  mat = res_t1_2Gy_vs_0Gy[, "stat", drop = FALSE],
  net = net,
  .source = "source",
  .target = "target",
  .mor = "mor"
)

sig_sample_acts_t1_2Gy <- sample_acts_t1_2Gy[sample_acts_t1_2Gy$p_value <= 0.05, ]


sample_acts_t1_4Gy <- run_ulm(
  mat = res_t1_4Gy_vs_0Gy[, "stat", drop = FALSE],
  net = net,
  .source = "source",
  .target = "target",
  .mor = "mor"
)

sig_sample_acts_t1_4Gy <- sample_acts_t1_4Gy[sample_acts_t1_4Gy$p_value <= 0.05, ]


sample_acts_t2_0.5Gy <- run_ulm(
  mat = res_t2_0.5Gy_vs_0Gy[, "stat", drop = FALSE],
  net = net,
  .source = "source",
  .target = "target",
  .mor = "mor"
)

sig_sample_acts_t2_0.5Gy <- sample_acts_t2_0.5Gy[sample_acts_t2_0.5Gy$p_value <= 0.05, ]


sample_acts_t2_1Gy <- run_ulm(
  mat = res_t2_1Gy_vs_0Gy[, "stat", drop = FALSE],
  net = net,
  .source = "source",
  .target = "target",
  .mor = "mor"
)


sig_sample_acts_t2_1Gy <- sample_acts_t2_1Gy[sample_acts_t2_1Gy$p_value <= 0.05, ]


sample_acts_t2_2Gy <- run_ulm(
  mat = res_t2_2Gy_vs_0Gy[, "stat", drop = FALSE],
  net = net,
  .source = "source",
  .target = "target",
  .mor = "mor"
)

sig_sample_acts_t2_2Gy <- sample_acts_t2_2Gy[sample_acts_t2_2Gy$p_value <= 0.05, ]


sample_acts_t2_4Gy <- run_ulm(
  mat = res_t2_4Gy_vs_0Gy[, "stat", drop = FALSE],
  net = net,
  .source = "source",
  .target = "target",
  .mor = "mor"
)

sig_sample_acts_t2_4Gy <- sample_acts_t2_4Gy[sample_acts_t2_4Gy$p_value <= 0.05, ]
```

## Filter top TFs in both signs

```{r}
n_tfs <- 25

f_sample_acts <- sample_acts_t1_0.5Gy %>%
  mutate(rnk = NA)

msk <- f_sample_acts$score > 0

f_sample_acts[msk, "rnk"] <- rank(-f_sample_acts[msk, "score"])

f_sample_acts[!msk, "rnk"] <- rank(-abs(f_sample_acts[!msk, "score"]))

tfs <- f_sample_acts %>%
  arrange(rnk) %>%
  head(n_tfs) %>%
  pull(source)

f_sample_acts <- f_sample_acts %>%
  filter(source %in% tfs)

# Plot
colors <- rev(RColorBrewer::brewer.pal(n = 11, name = "RdBu")[c(2, 10)])

ggplot(f_sample_acts, aes(x = reorder(source, score), y = score)) +
  geom_bar(aes(fill = score), stat = "identity") +
  scale_fill_gradient2(
    low = "darkblue", high = "indianred",
    mid = "whitesmoke", midpoint = 0
  ) +
  theme_minimal() +
  theme(
    axis.title = element_text(face = "bold", size = 12),
    axis.text.x =
      element_text(angle = 45, hjust = 1, size = 10, face = "bold"),
    axis.text.y = element_text(size = 10, face = "bold"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  xlab("TFs")
```

## Decoupler pheatmap

```{r}
sample_acts <- run_ulm(
  mat = decoupler_counts,
  net = net,
  .source = "source",
  .target = "target",
  .mor = "mor",
  minsize = 5
)

# Transform to wide matrix
sample_acts_mat <- sample_acts %>%
  pivot_wider(
    id_cols = "condition", names_from = "source",
    values_from = "score"
  ) %>%
  column_to_rownames("condition") %>%
  as.matrix()

# Get top tfs with more variable means across clusters
tfs <- sample_acts %>%
  group_by(source) %>%
  summarise(std = sd(score)) %>%
  arrange(-abs(std)) %>%
  head(30) %>%
  pull(source)
sample_acts_mat <- sample_acts_mat[, tfs]

# Scale per sample
sample_acts_mat <- scale(sample_acts_mat)

# Choose color palette
palette_length <- 100
my_color <- colorRampPalette(c("Darkblue", "white", "red"))(palette_length)

my_breaks <- c(
  seq(-3, 0, length.out = ceiling(palette_length / 2) + 1),
  seq(0.05, 3, length.out = floor(palette_length / 2))
)

# column annotation

decoupler_column_annotation <- data.frame(Time_Point = time_point, Sex = sex, Dose = dose)

# Plot
pheatmap(t(sample_acts_mat), border_color = NA, color = my_color, breaks = my_breaks, annotation_col = column_annotation, cluster_cols = FALSE)
```

#Deconvolution-- immune cells relative abundance

```{r}
# QuantiseqR

tpm_mat <- assay(dds, "abundance")

tpm_mat <- tpm_mat[rowSums(tpm_mat) != 0, ]

tpm_mat <- tpm_mat[!rownames(tpm_mat) %in% rownames(ddd_low_var), ]


tpm_genes <- rownames(tpm_mat)

tpm_genes <- mapIds(
  org.Hs.eg.db,
  keys = tpm_genes,
  column = "SYMBOL",
  keytype = "ENSEMBL",
  multiVals = "first"
)

rownames(tpm_mat) <- tpm_genes

# running quantiseqr

deconv_quantiseq <- quantiseqr::run_quantiseq(
  expression_data = tpm_mat,
  signature_matrix = "TIL10",
  is_arraydata = FALSE,
  is_tumordata = TRUE,
  scale_mRNA = TRUE
)

quantiplot(deconv_quantiseq)

pheatmap(t(deconv_quantiseq[, -1]) - rowMeans(t(deconv_quantiseq[, -1])), cluster_cols = TRUE)


# immunedeconv with Abis method

res_abis <- immunedeconv::deconvolute(gene_expression = tpm_mat, method = "abis")

res_abis_matrix <- as.matrix(res_abis)

rownames(res_abis_matrix) <- res_abis_matrix[, 1]

res_abis_matrix <- res_abis_matrix[, -1]

rows_res_abis <- rownames(res_abis_matrix)

res_abis_matrix <- apply(res_abis_matrix, 2, as.numeric)

rownames(res_abis_matrix) <- rows_res_abis

pheatmap(res_abis_matrix_ordered - rowMeans(res_abis_matrix_ordered), cluster_cols = FALSE)

# Omnideconv

# the bulk matrix is the dds_mat prepared above

# prepare the single cell matrix

is_outlier <- function(SeuratObject, metric, nmads) {
  eval(parse(text = paste0("M <- SeuratObject$", metric)))
  outlier <- (M < median(M) - nmads * mad(M)) | (M > median(M) + nmads * mad(M))
  return(outlier)
}


mito_genes <- grep("^MT-", rownames(seurat_obj@assays$SCT@counts), value = TRUE)

seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, features = mito.genes)

# Filtering

check_outliers_nFeature <- is_outlier(seurat_obj, "nFeature_RNA", 5)
check_outliers_nCount <- is_outlier(seurat_obj, "nCount_RNA", 5)
check_outliers_mito <- is_outlier(seurat_obj, "percent.mt", 3)


non_outliers_nFeature <- names(check_outliers_nFeature[!check_outliers_nFeature])
non_outliers_nCount <- names(check_outliers_nCount[!check_outliers_nCount])
non_outliers_mito <- names(check_outliers_mito[!check_outliers_mito])

non_outliers <- intersect(non_outliers_nFeature, non_outliers_nCount) %>% intersect(non_outliers_mito)

seurat_obj <- subset(seurat_obj, cells = non_outliers)


View(as.data.frame(table(seurat_obj$free_annotation, dnn = list("celltype_major")), responseName = "number_cells"))



sampled.metadata <- seurat_obj@meta.data %>%
  rownames_to_column(., "barcode") %>%
  group_by(., free_annotation) %>%
  nest() %>%
  dplyr::mutate(n = map_dbl(data, nrow)) %>%
  dplyr::mutate(n = min(n, 500)) %>%
  ungroup() %>%
  dplyr::mutate(samp = map2(data, n, sample_n)) %>%
  dplyr::select(-data) %>%
  unnest(samp)

single.cell.data.sampled <- subset(seurat_obj, cells = sampled.metadata$barcode)

sampled.metadata <- sampled.metadata[!sampled.metadata$free_annotation %in% c(
  "erythrocyte",
  "granulocyte",
  "hematopoietic stem cell",
  "myeloid progenitor",
  "non-classical monocyte",
  "plasmablast",
  "plasmacytoid dendritic cell"
), ]


count_matrix <- GetAssayData(single.cell.data.sampled, layer = "counts")


# running omnideconv-bayesprism

deconv_bayesprism <- deconvolute(
  bulk_gene_expression = dds_mat,
  single_cell_object = sc_blood_modified,
  cell_type_annotations = Blood_annotation_modified,
  method = "bayesprism",
  n_cores = 12
)

omnideconv::plot_deconvolution(list("bayesprism" = deconv_bayesprism), "bar", "method", "Spectral")

pheatmap(t(deconv_bayesprism) - rowMeans(t(deconv_bayesprism)))

# CIBERSORTx (in web-based tool) was also performed
```

# Exporting the data

```{r}
exportr <- function(resuSet, resu_name = "out_SET") {
  message("Exporting the whole set for ", resu_name)
  for (i in names(resuSet)) {
    message("    ", resu_name, "  ---  ", i)
    curset <- resuSet[[i]]

    if (!is.null(curset$tbl_res_all)) {
      message("        Exporting results from DESeq... ")
      export_deseq_file <- paste0(resu_name, "_", i, "_tbl_res_DESeq.xlsx")
      writexl::write_xlsx(curset$tbl_res_all, export_deseq_file)
      message("          ", export_deseq_file)
    }

    if (!is.null(curset$clupro_tbl)) {
      message("        Exporting results from clusterProfiler... ")
      export_clupro_file <- paste0(resu_name, "_", i, "_tbl_cluPro.xlsx")
      writexl::write_xlsx(curset$clupro_tbl@result, export_clupro_file)
      message("          ", export_clupro_file)
    }
  }
}

exportr(resuSet = resuSet, resu_name = "Phybion")
```

# Session information

```{r sessioninfo}
sessionInfo()
```
